File: bindings/wasm/src/modular.ts
/**
 * Copyright (c) 2026 Edison Manrique Chocce
 * Licencia Dual: AGPL-3.0 | Licencia Comercial para uso propietario.
 *
 * ERMC WASM — Aritmética Modular en F_p (v0.4.0)
 */

import { getWasmExports } from "./wasm";

const U64_MAX = BigInt("18446744073709551615");

export interface WasmModularLib {
  add(a: bigint, b: bigint, p: bigint): bigint;
  sub(a: bigint, b: bigint, p: bigint): bigint;
  mul(a: bigint, b: bigint, p: bigint): bigint;
  pow(base: bigint, exp: bigint, p: bigint): bigint;
  inverse(a: bigint, p: bigint): bigint | null;
  sqrt(a: bigint, p: bigint): bigint | null;
  isQR(a: bigint, p: bigint): boolean;
}

export function getWasmModular(): WasmModularLib {
  const exp = getWasmExports();
  return {
    add: (a, b, p) => exp.ermc_mod_add(a, b, p),
    sub: (a, b, p) => exp.ermc_mod_sub(a, b, p),
    mul: (a, b, p) => exp.ermc_mod_mul(a, b, p),
    pow: (base, e, p) => exp.ermc_mod_pow(base, e, p),
    inverse(a, p) {
      const r = exp.ermc_mod_inverse(a, p);
      return r === U64_MAX ? null : r;
    },
    sqrt(a, p) {
      const r = exp.ermc_mod_sqrt(a, p);
      return r === U64_MAX ? null : r;
    },
    isQR(a, p) {
      const e = (p - 1n) / 2n;
      return exp.ermc_mod_pow(a, e, p) === 1n;
    },
  };
}

export { getWasmModular as getModular };

File: bindings/python/ermc/modular.py
# Copyright (c) 2026 Edison Manrique Chocce
# Todos los derechos reservados.
# Licencia Dual: AGPL-3.0 para uso académico/personal | Licencia Comercial requerida para uso propietario.
# Ver archivo LICENSE en la raíz del proyecto para términos completos.

from typing import Optional
from .ffi import get_lib

U64_MAX = 18446744073709551615


class Modular:
    """Aritmética de campos finitos F_p (enteros de 64 bits sin signo)."""

    @staticmethod
    def add(a: int, b: int, p: int) -> int:
        """Suma modular: (a + b) mod p"""
        return get_lib().ermc_mod_add(int(a), int(b), int(p))

    @staticmethod
    def sub(a: int, b: int, p: int) -> int:
        """Resta modular: (a - b) mod p"""
        return get_lib().ermc_mod_sub(int(a), int(b), int(p))

    @staticmethod
    def mul(a: int, b: int, p: int) -> int:
        """Multiplicación modular: (a * b) mod p"""
        return get_lib().ermc_mod_mul(int(a), int(b), int(p))

    @staticmethod
    def pow(base: int, exp: int, p: int) -> int:
        """Exponenciación modular binaria: base^exp mod p"""
        return get_lib().ermc_mod_pow(int(base), int(exp), int(p))

    @staticmethod
    def inverse(a: int, p: int) -> Optional[int]:
        """Inverso modular a^{-1} mod p (Algoritmo Extendido de Euclides), o None si no existe."""
        res = get_lib().ermc_mod_inverse(int(a), int(p))
        return None if res == U64_MAX else res

    @staticmethod
    def sqrt(a: int, p: int) -> Optional[int]:
        """Raíz cuadrada modular √a mod p (Algoritmo Tonelli-Shanks), o None si no es residuo cuadrático."""
        res = get_lib().ermc_mod_sqrt(int(a), int(p))
        return None if res == U64_MAX else res

    @staticmethod
    def is_qr(a: int, p: int) -> bool:
        """Criterio de Euler: verifica si a es residuo cuadrático mod p."""
        e = (int(p) - 1) // 2
        return Modular.pow(a, e, p) == 1
